model.py

```diff
@@ -239,3 +314,185 @@ def forward(self, x):
+
+            # output is nb_frames, nb_samples, nb_channels, nb_bins
+            if spectrograms is None:
+                # allocate the spectrograms variable
+                spectrograms = torch.zeros(
+                    target_spectrogram.shape + (nb_sources,),
+                    dtype=torch.float64,
+                    device=target_spectrogram.device
+                )
+
+            spectrograms[..., j] = target_spectrogram
+
+        # transposing it as
+        # (nb_samples, nb_frames, nb_bins,{1,nb_channels}, nb_sources)
+        spectrograms = spectrograms.permute(1, 0, 3, 2, 4).to(torch.float64)
+
+        # getting the STFT of mix:
+        # (nb_samples, nb_channels, nb_bins, nb_frames, 2)
+        mix_stft = unmix_target.stft(audio).to(torch.float64)
+
+        # rearranging it into:
+        # (nb_samples, nb_frames, nb_bins, nb_channels, 2) to feed
+        # into filtering methods
+        mix_stft = mix_stft.permute(0, 3, 2, 1, 4)
+
+        # create an additional target if we need to build a residual
+        targets = list(self.targets.keys())
+        if self.residual is not None:
+            targets += [self.residual]
+            nb_sources += 1
+
+        if len(targets) == 1 and self.niter > 0:
+            raise Exception('Cannot use EM if only one target is estimated.'
+                            'Provide two targets or create an additional '
+                            'one with `--residual`')
+
+        nb_frames = spectrograms.shape[1]
+        targets_stft = torch.zeros(
+            mix_stft.shape + (nb_sources, ),
+            dtype=torch.float32,
+            device=mix_stft.device
+        )
+        for sample in range(nb_samples):
+            pos = 0
+            batch_size = self.batch_size if self.batch_size else nb_frames
+            while pos < nb_frames:
+                t = torch.arange(pos, min(nb_frames, pos+batch_size))
+                pos = t[-1] + 1
+
+                targets_stft[sample, t] = wiener(
+                    spectrograms[sample, t], mix_stft[sample, t],
+                    self.niter, use_softmask=False,
+                    residual=self.residual
+                ).to(torch.float32)
+        estimates = {}
+
+        # getting to (nb_samples, channel, fft_size, n_frames, 2, nb_sources)
+        targets_stft = targets_stft.permute(0, 3, 2, 1, 4, 5)
+
+        # Now performing the inverse STFTs
+        for j, name in enumerate(targets):
+            estimates[name] = torch.cat([
+                istft(
+                    targets_stft[sample, ..., j],
+                    n_fft=unmix_target.stft.n_fft,
+                    hop_length=unmix_target.stft.n_hop,
+                    window=unmix_target.stft.window,
+                    center=unmix_target.stft.center,
+                    normalized=False,
+                    onesided=True,
+                    pad_mode='reflect',
+                    length=audio.shape[-1]
+                ).transpose(0, 1)[None, ...]
```

Comment:
torchaudio's `istft` supports `(…, fft_size, n_frame, 2)` so we might be faster with stacking all estimates into a tensor with `nb_targets` as first dimension, doing one big istft after...

Another reason we should maybe not use dict outputs?

Reply:
if we don't have EM iterations, we do each istft one after the other because it's more ram efficient (we don't need to store all the complex STFT)

Reply:
but you already have them in `target_stft`, no?

Reply:
implemented